File: 1.30/solution.rb
module SumIter
  def self.calc(a, b, term, nexta)
    calc_iter(a, b, term, nexta, 0)
  end

  def self.calc_iter(a, b, term, nexta, sum)
    return sum if a > b
    calc_iter(nexta.call(a), b, term, nexta, sum + term.call(a))
  end
end

itself = -> (x) { x }
inc = -> (x) { x.next }

puts "Sum 1..20: #{SumIter.calc(1, 20, itself, inc)}"

File: 1.30/solution.exs
defmodule SumIter do
  def calc(a, b, term, nexta) do
    calc_iter(a, b, term, nexta, 0)
  end

  def calc_iter(a, b, _term, _nexta, sum) when a > b, do: sum

  def calc_iter(a, b, term, nexta, sum) do
    calc_iter(nexta.(a), b, term, nexta, sum + term.(a))
  end
end

itself = fn(x) -> x end
inc = fn (x) -> x + 1 end

IO.puts "Sum 1..20: #{SumIter.calc(1, 20, itself, inc)}"